An unknown error occurred: couldn't rename path

After performing a rustup today, the build for libpnet has stopped working:

``` bash
$ export VERBOSE=1
$ rustc -v
rustc 0.13.0-nightly (bd7138dd6 2014-10-27 23:02:55 +0000)
$ cargo -V
cargo 0.0.1-pre-nightly (57684d0 2014-10-27 23:49:37 +0000)
$ git clone git@github.com:libpnet/libpnet.git
Cloning into 'libpnet'...
remote: Counting objects: 190, done.
remote: Compressing objects: 100% (103/103), done.
remote: Total 190 (delta 101), reused 162 (delta 86)
Receiving objects: 100% (190/190), 78.35 KiB | 0 bytes/s, done.
Resolving deltas: 100% (101/101), done.
Checking connectivity... done
$ cd libpnet/
$ make
bash build.sh
++ which rustc
+ RUSTC=/usr/local/bin/rustc
++ which rustdoc
+ RUSTDOC=/usr/local/bin/rustdoc
++ which cargo
+ CARGO=/usr/local/bin/cargo
++ which sudo
+ SUDO=/usr/bin/sudo
++ uname -s
+ SYSTEM=Linux
+ TESTER='/usr/local/bin/cargo test'
++ which clang
+ CC=/usr/bin/clang
+ CARGO_FLAGS=
++ ifconfig
++ perl -pe '/^[A-z0-9]+:([^\n]|\n\t)*status: active/'
++ cut -f1 -d:
++ egrep 'UP| active'
++ grep active -B1
++ head -n1
+ PNET_TEST_IFACE=
+ mkdir -p target/doc
+ mkdir -p target/benches
+ [[ 1 == \1 ]]
+ CARGO_FLAGS=--verbose
+ TESTER='/usr/local/bin/cargo test --verbose'
+ [[ ! -x /usr/local/bin/cargo ]]
+ case "$1" in
+ build
+ [[ -x /usr/local/bin/cargo ]]
+ /usr/local/bin/cargo build --verbose --release
   Compiling pnet v0.0.1 (file:///tmp/libpnet)
     Running `rustc /tmp/libpnet/src/lib.rs --crate-name pnet --crate-type rlib --crate-type dylib --opt-level 3 --cfg ndebug -C metadata=4ef40c18c7afd0de -C extra-filename=-4ef40c18c7afd0de --out-dir /tmp/libpnet/target/release --dep-info /tmp/libpnet/target/release/.fingerprint/pnet-4ef40c18c7afd0de/dep-lib-pnet -L /tmp/libpnet/target/release -L /tmp/libpnet/target/release/deps`
$ make test
bash build.sh test
++ which rustc
+ RUSTC=/usr/local/bin/rustc
++ which rustdoc
+ RUSTDOC=/usr/local/bin/rustdoc
++ which cargo
+ CARGO=/usr/local/bin/cargo
++ which sudo
+ SUDO=/usr/bin/sudo
++ uname -s
+ SYSTEM=Linux
+ TESTER='/usr/local/bin/cargo test'
++ which clang
+ CC=/usr/bin/clang
+ CARGO_FLAGS=
++ ifconfig
++ egrep 'UP| active'
++ head -n1
++ cut -f1 -d:
++ grep active -B1
++ perl -pe '/^[A-z0-9]+:([^\n]|\n\t)*status: active/'
+ PNET_TEST_IFACE=
+ mkdir -p target/doc
+ mkdir -p target/benches
+ [[ 1 == \1 ]]
+ CARGO_FLAGS=--verbose
+ TESTER='/usr/local/bin/cargo test --verbose'
+ [[ ! -x /usr/local/bin/cargo ]]
+ case "$1" in
+ run_test
+ build_test
+ [[ -x /usr/local/bin/cargo ]]
+ /usr/local/bin/cargo test --no-run --verbose
   Compiling pnet v0.0.1 (file:///tmp/libpnet)
     Running `rustc /tmp/libpnet/src/lib.rs --crate-name pnet --crate-type rlib --crate-type dylib -g -C metadata=4ef40c18c7afd0de -C extra-filename=-4ef40c18c7afd0de --out-dir /tmp/libpnet/target --dep-info /tmp/libpnet/target/.fingerprint/pnet-4ef40c18c7afd0de/dep-lib-pnet -L /tmp/libpnet/target -L /tmp/libpnet/target/deps`
     Running `rustc /tmp/libpnet/src/lib.rs --crate-name pnet --crate-type rlib --crate-type dylib -g --test -C metadata=18776064f6b66fc4 -C extra-filename=-18776064f6b66fc4 --out-dir /tmp/libpnet/target --dep-info /tmp/libpnet/target/.fingerprint/pnet-4ef40c18c7afd0de/dep-test-lib-pnet -L /tmp/libpnet/target -L /tmp/libpnet/target/deps`
     Running `rustc /tmp/libpnet/examples/transport_echo_server.rs --crate-name transport_echo_server --crate-type bin -g --out-dir /tmp/libpnet/target/examples --dep-info /tmp/libpnet/target/.fingerprint/pnet-4ef40c18c7afd0de/dep-bin-transport_echo_server -L /tmp/libpnet/target -L /tmp/libpnet/target/deps`
     Running `rustc /tmp/libpnet/examples/packetdump.rs --crate-name packetdump --crate-type bin -g --out-dir /tmp/libpnet/target/examples --dep-info /tmp/libpnet/target/.fingerprint/pnet-4ef40c18c7afd0de/dep-bin-packetdump -L /tmp/libpnet/target -L /tmp/libpnet/target/deps`
+ echo 'Setting permissions for test suite - enter sudo password if prompted'
Setting permissions for test suite - enter sudo password if prompted
+ case "$SYSTEM" in
++ id -u
+ [[ 1001 != \0 ]]
+ /usr/bin/sudo setcap cap_net_raw+ep target/pnet-18776064f6b66fc4
+ RUST_TEST_TASKS=1
+ /usr/local/bin/cargo test --verbose
       Fresh pnet v0.0.1 (file:///tmp/libpnet)
Build failed, waiting for other jobs to finish...
An unknown error occurred
couldn't rename path (no such file or directory (No such file or directory); from=/tmp/libpnet/target/old-examples/examples/transport_echo_server; to=/tmp/libpnet/target/examples/examples/transport_echo_server)
make: *** [test] Error 101
```

It should be possible to reproduce using (the build script needs root access, but the issue can be reproduced without):

``` bash
$ cargo test --no-run --verbose
$ cargo test --verbose
```

The issue doesn't seem to occur without the `--no-run` step.

--- a/src/cargo/ops/cargo_rustc/context.rs
+++ b/src/cargo/ops/cargo_rustc/context.rs
@@ -225,9 +225,8 @@ impl<'a, 'b> Context<'a, 'b> {
         let stem = target.file_stem();
 
         let mut ret = Vec::new();
-        if target.is_example() {
-            ret.push(format!("examples/{}{}", stem, self.target_exe));
-        } else if target.is_bin() || target.get_profile().is_test() {
+        if target.is_example() || target.is_bin() ||
+           target.get_profile().is_test() {
             ret.push(format!("{}{}", stem, self.target_exe));
         } else {
             if target.is_dylib() {

--- a/tests/test_cargo_test.rs
+++ b/tests/test_cargo_test.rs
@@ -1254,3 +1254,24 @@ test!(example_bin_same_name {
     assert_that(p.process(p.bin("examples/foo")),
                 execs().with_status(0).with_stdout("example\n"));
 })
+
+test!(test_with_example_twice {
+    let p = project("foo")
+        .file("Cargo.toml", r#"
+            [package]
+            name = "foo"
+            version = "0.0.1"
+            authors = []
+        "#)
+        .file("src/bin/foo.rs", r#"fn main() { println!("bin"); }"#)
+        .file("examples/foo.rs", r#"fn main() { println!("example"); }"#);
+
+    println!("first");
+    assert_that(p.cargo_process("test").arg("-v"),
+                execs().with_status(0));
+    assert_that(&p.bin("examples/foo"), existing_file());
+    println!("second");
+    assert_that(p.process(cargo_dir().join("cargo")).arg("test").arg("-v"),
+                execs().with_status(0));
+    assert_that(&p.bin("examples/foo"), existing_file());
+})

